SourceCode/CRTRenderer.cpp

```diff
@@ -25,68 +13,130 @@ void CRTRenderer::renderScene(const char* outputname) const
     CRTImageSaver::saveImage(outputname, img);
 }
 
-CRTImage CRTRenderer::renderScene() const {
-    return renderScene(visualizeByDepth);
+void CRTRenderer::renderSceneBarycentic(const char* outputname) const
+{
+    CRTImage img = renderSceneBarycentic();
+    CRTImageSaver::saveImage(outputname, img);
 }
 
-float CRTRenderer::getMaxDistanceToCamera(const std::vector<CRTTriangle>& triangles) const {
-    float maxDistance = -1;
-    for (int i = 0; i < triangles.size(); i++) {
-        maxDistance = std::max(maxDistance, std::abs(triangles[i].distanceToPoint(scene->getCamera().getPosition())));
+CRTImage CRTRenderer::renderScene() const
+{
+    const unsigned imageHeight = scene->getSettings().imageSettings.height;
+    const unsigned imageWidth = scene->getSettings().imageSettings.width;
+
+    CRTImage image(imageHeight, std::vector<CRTColor>(imageWidth, scene->getSettings().bgColor));
+
+    for (int rowId = 0; rowId < imageHeight; rowId++) {
+        for (int colId = 0; colId < imageWidth; colId++) {
+            CRTRay ray = scene->getCamera().getRayForPixel(rowId, colId); // we generate the ray "on demand"
+            image[rowId][colId] = rayTrace(ray, 0, CRTVector(1, 1, 1)); // start from depth 0 and standard albedo values
+        }
     }
-    return maxDistance;
+    return image;
 }
 
-CRTImage CRTRenderer::renderScene(std::function<CRTColor(const CRTVector&, const CRTTriangle&, float, const CRTVector&)> visualization) const 
+CRTColor CRTRenderer::rayTrace(const CRTRay& ray, int depth, const CRTVector& addedAlbedo) const
+{
+    CRTVector closestHit;
+    CRTVector normalVector;
+    const CRTMesh* meshHit = nullptr;
+    float minDistanceToCamera = FLT_MAX;
+    bool intersects = false;
+    for (int i = 0; i < scene->getGeometryObjects().size(); i++) {
+        std::tuple<bool, CRTVector, CRTVector> hit = scene->getGeometryObjects()[i].intersectsRay(ray);
+        if (std::get<0>(hit)) {
+            intersects = true;
+            double distanceToCamera = (std::get<1>(hit) - ray.getOrigin()).length();
+            if (distanceToCamera < minDistanceToCamera) {
+                minDistanceToCamera = distanceToCamera;
+                closestHit = std::get<1>(hit);
+                normalVector = std::get<2>(hit);
+                meshHit = &scene->getGeometryObjects()[i];
+            }
+        }
+    }
+    if (intersects && meshHit) {
+        const CRTVector& reflectiveAlbedo = meshHit->getMaterial().albedo;
+        CRTVector newAlbedo(addedAlbedo.x * reflectiveAlbedo.x, addedAlbedo.y * reflectiveAlbedo.y, addedAlbedo.z * reflectiveAlbedo.z);
```

Comment:
You have to just backpropagate the color on the light path, without multiplying it.